[WIP] Implemented nearest neighborhood index service

aashish24 opened the pull request.

aashish24 commented:
@Purg please have a look and let me know if something is wrong or missing.

Purg commented on docs/examples/caffe_build_index.md:
```diff
@@ -0,0 +1,54 @@
+The following is a concrete example of performing nearest neiborhood computation
+using a set of ten butterfly images. It assumes you have set up the
+caffe (vrc2) environment (http://caffe.berkeleyvision.org/install_apt.html)
```
This descriptor implementation also works on master (what I'm working with at the moment). The other Caffe descriptor that uses the external c++ executable requires rc2, however.

aashish24 replied:
when I tried to use master of caffe, I ran into a build issue but I will update here saying that it should work with caffe master as well.

Purg commented on docs/examples/caffe_build_index.md:
```diff
@@ -0,0 +1,54 @@
+        network_is_bgr=True,
+        data_layer="data",
+        load_truncated_images=True)
+# cd.generate_model(el)
```
going to want to remove this commented-out line in relation with the specific executable being used. It might be confusing to people.

aashish24 replied:
Ah yes, sorry that was a left over. I will take it out.

Purg commented on python/smqtk/web/nn_descriptor_service/__init__.py:
```diff
@@ -0,0 +1,231 @@
+
+        # Cache of DescriptorGenerator instances so we don't have to continuously
+        # initialize them as we get requests.
+        self.descriptor_cache = {}
```
This and its lock are now no longer used anywhere.

aashish24 replied:
Good catch..will remove it.

Purg commented on python/smqtk/web/nn_descriptor_service/__init__.py:
```diff
@@ -0,0 +1,231 @@
+        @self.route("/nn/<path:uri>")
+        def compute_nearest_neighbors(uri):
+            """
+            # Data modes for upload/use
```
We're going to want to switch this over to reStructured text format (see the descriptor_service method comments) since the addition of sphinx documentation generation.

aashish24 replied:
Thanks, will update the document in my next commit.

Purg commented on python/smqtk/web/nn_descriptor_service/__init__.py:
```diff
@@ -0,0 +1,231 @@
+                    message = "Data content type issue: %s" % str(ex)
+
+            # fail here if de is None
+            # Default is 8
```
Going to want some protection here if descriptor generation failed for some reason.

Purg commented on python/smqtk/web/nn_descriptor_service/__init__.py:
```diff
@@ -0,0 +1,231 @@
+
+        return de
+
+APPLICATION_CLASS = NearestNeighborServiceServer
```
Technically shouldn't need this any more with recent (ish) plugin discovery updates, but its fine the way it is (check out the get_plugins doc-strings).

Purg commented on python/smqtk/web/nn_descriptor_service/__init__.py:
```diff
@@ -0,0 +1,231 @@
+            num_neighbors = flask.request.args.get("num_neighbors", 8)
+            neighbors, _ = self.nn_index.nn(descriptor, n=num_neighbors)
+
+            # TODO: Return the optional descriptor vector for the neighbors
```
This remaining TODO if you have time. Could just be another URL arg (boolean).

Purg commented:
Code examples were apparently moved to ``docs/examples.rst`` (single file). The directory your current examples file is would be removed in #170 from @kfieldho. We will either want to move your example to that ``docs/examples.rst`` file, or we should revise the current examples structure.

aashish24 commented:
@Purg hopefully I addressed all of comments except the example one. May be if the changes looks good to you, we can merge this branch and have @kfieldho update his PR?

Thanks

Purg commented on python/smqtk/web/nn_descriptor_service/__init__.py:
```diff
@@ -0,0 +1,230 @@
+                        compute_descriptor(de, self.descr_elem_factory)
+
+                    if descriptor is not None:
+                        descriptor.set_vector(descriptor.vector().flatten())
```
What is this flattening for? Descriptors should always be 1D. Were you seeing a case where they weren't?

aashish24 replied:
Right and that was the TODO I had earlier which I removed as I thought you looked at it. Yes, I was getting (4096, 1, 1) both when building and using the index and since flann was not liking it. I debug further but I could not find any solutions to the problem other than flattening it. You can see in my build index example, I am doing the same.

Purg replied:
Let me run your example and see what I get. Descriptors should be being output in 1 dimension, so if its reproducible, something might be going wrong with the descriptor generator.

aashish24 replied:
> Let me run your example and see what I get. Descriptors should be being output in 1 dimension, so if its reproducible, something might be going wrong with the descriptor generator.

Cool thanks @Purg. Also, I added the guards so please have a look again.

Purg commented on python/smqtk/web/nn_descriptor_service/__init__.py:
```diff
@@ -0,0 +1,230 @@
+
+            neighbors = []
+            if descriptor is not None:
+                neighbors, _ = self.nn_index.nn(descriptor, n=num_neighbors)
```
This could raise ``ValueError`` exceptions (if descriptor didn't have a vector, or if the index didn't contain anything), so we're going to want to protect against that.

aashish24 replied:
I see. I will add try catch, thanks for the info.

aashish24 commented:
@Purg I updated the code as discussed and verified the results

Purg commented on python/smqtk/algorithms/descriptor_generator/caffe_descriptor.py:
```diff
@@ -408,8 +408,9 @@ def _process_batch(self, uuids4proc, data_elements, descr_elements, procs):
-
+            # In case caffe generates multidimensional array (rows, 1, 1)
+            # or matrix format result
+            descr_elements[uid].set_vector(numpy.array(v, copy=False).flatten())
```
Turns out I was wrong again. ``flatten`` is going to always make a copy, which might not be the best idea. Might end up being better to check ``v``'s shape first to decide whether flattening is necessary:
```
...
if v.ndim > 1:
    v = numpy.array(v, copy=False).flatten()
descr_elements[uid].set_vector(v)
...
```

aashish24 replied:
> if v.ndim > 1:
>     v = numpy.array(v, copy=False).flatten()
> descr_elements[uid].set_vector(v)
> ...

if that's the case, then does it makes sense to pass copy=False? Let me look at the docs

Purg commented:
Oh, yeah. Started maintaining release notes files in ``docs/release_nodes/``. Please update the ``since-last-release.md`` file with notes pertaining to features added/updated and bugs fixed.

aashish24 commented:
@Purg would it be okay if I do that in next PR?

Purg commented:
The next PR being...

aashish24 commented:
> The next PR being...

NM. I was thinking of cleaning up some documentation stuff. Should be ready for PR :smile: now

aashish24 commented:
> Add this under a Web Tools section. This would be more in-line with the precedent set in other release notes files.

How about Web / Services section? I just forced push. Have a look

Purg commented:
Yeah, that's fine. I'll go over this once more in the morning before merging (heading out right now).

Purg commented:
Given that examples will currently live in ``docs/examples.rst``, move your example doc into that.

aashish24 commented:
@Purg that branch is not merged into master yet, once this one goes in I can update that branch. I think we should discuss the example structure a bit more. Can we merge this branch?
